### ISSUE TITLE
Cookies not attached on request to local server

### ISSUE THREAD

**Issue by kashaCode** [peak3d/inputstream.adaptive#1557]
### Describe the bug
I am running a local proxy (basically a service addon + proxy written in Python using simplehttpserver) to fix some things in the manifest and then passing off to ISA (Setup looks like: Server <--> My proxy <---> ISA). Requests from ISA are passed as it is to remote server, and response from server are passed as it is to ISA, with only some trivial changes in headers and fixing some things in the manifest. The remote server issues updated cookies in each response, which must be attached to the next request. 

ISA when fetching manifest directly from remote server (without going through my proxy) correctly attaches cookies from each server response to next request made to server. However, when I change the manifest URL to point to my local proxy, ISA fails to attach cookies from previous response to the next request. 

## Actual Behavior
I have added the server response and ISA's request below. As you can see, the response issued by remote server and my proxy is nearly identical (with only trivial differences), yet the behaviour of ISA is different in both cases. **My theory is that it attaches cookies when making requests to remote HTTPS server, but not when making requests to local HTTP server (`http://127.0.0.1:port/path.....`).**

### Behaviour when connecting with remote server
Here is the remote server response for initial request, which contains cookie:

```
HTTP/1.1 200 OK
Connection: keep-alive
Content-Encoding: gzip
Content-Length: 2457
x-amz-id-2: XM2fxVrs1S+0H/O2BFvy6MmmSjVOLcSmV0EqB2vKIzxxahqrTcClSYJI3gmaCYyZ752DH/5D2Dc=
x-amz-request-id: CWDHN0R93EF1QB5Y
Last-Modified: Tue, 31 Oct 2023 16:53:03 GMT
ETag: "2ae53a921d2115602f403165c4068626"
x-amz-server-side-encryption: AES256
Content-Type: binary/octet-stream
Server: AmazonS3
Accept-Ranges: bytes
Date: Fri, 17 May 2024 03:21:50 GMT
Via: 1.1 varnish
Age: 33710
X-Served-By: cache-rj-vabb4300025-RJ-VABB
X-Cache: HIT
X-Cache-Hits: 2
X-Timer: S1715916110.219052,VS0,VE0
Vary: Accept-Encoding
Access-Control-Allow-Methods: GET,HEAD,OPTIONS
Access-Control-Allow-Credentials: true
access-control-max-age: 3600
access-control-expose-headers: *
access-control-allow-headers: *
Set-Cookie: __hdnea__=st=1715916110~exp=1716002510~acl=/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/*~hmac=deb7ef23ebf9112155f8055630341f95a7c5f9baa6b27c508d89960d67097d90; Domain=gec-007-jc.slivcdn.com; path=/; Expires=Sat, 18 May 2024 03:21:50 GMT; SameSite=None; Secure

----------------------  Response unzipped   ----------------------

#EXTM3U
#EXT-X-VERSION:4
#EXT-X-INDEPENDENT-SEGMENTS
#EXT-X-STREAM-INF:BANDWIDTH=389104,AVERAGE-BANDWIDTH=237109,CODECS="avc1.4d400c,mp4a.40.2",RESOLUTION=256x144,FRAME-RATE=25.000,AUDIO="program_audio"
1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_200k.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=525216,AVERAGE-BANDWIDTH=343045,CODECS="avc1.4d4015,mp4a.40.2",RESOLUTION=426x240,FRAME-RATE=25.000,AUDIO="program_audio"
1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_300k.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=814736,AVERAGE-BANDWIDTH=526284,CODECS="avc1.77.30,mp4a.40.2",RESOLUTION=640x360,FRAME-RATE=25.000,AUDIO="program_audio"
1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_500K.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=1192992,AVERAGE-BANDWIDTH=723497,CODECS="avc1.77.30,mp4a.40.2",RESOLUTION=854x480,FRAME-RATE=25.000,AUDIO="program_audio"
1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_800K.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=1713000,AVERAGE-BANDWIDTH=1133280,CODECS="avc1.4d401f,mp4a.40.2",RESOLUTION=960x540,FRAME-RATE=25.000,AUDIO="program_audio"
1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_1200K.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=2696240,AVERAGE-BANDWIDTH=1355768,CODECS="avc1.4d401f,mp4a.40.2",RESOLUTION=1280x720,FRAME-RATE=25.000,AUDIO="program_audio"
1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_2000k.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=3847552,AVERAGE-BANDWIDTH=1783712,CODECS="avc1.4d401f,mp4a.40.2",RESOLUTION=1280x720,FRAME-RATE=25.000,AUDIO="program_audio"
1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_3000k.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=5452696,AVERAGE-BANDWIDTH=3921845,CODECS="avc1.4d4028,mp4a.40.2",RESOLUTION=1920x1080,FRAME-RATE=25.000,AUDIO="program_audio"
1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_4200k.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=7051448,AVERAGE-BANDWIDTH=4278185,CODECS="avc1.4d4028,mp4a.40.2",RESOLUTION=1920x1080,FRAME-RATE=25.000,AUDIO="program_audio"
1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_5500k.m3u8
#EXT-X-MEDIA:TYPE=AUDIO,GROUP-ID="program_audio",LANGUAGE="hin",NAME="Hindi",AUTOSELECT=YES,DEFAULT=YES,URI="1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_Hindi96k.m3u8"

```

ISA makes next request correctly which includes Cookie from previous response:

```
GET https://gec-007-jc.slivcdn.com/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_5500k.m3u8 HTTP/1.1
Host: gec-007-jc.slivcdn.com
User-Agent: plaYtv/7.1.3 (Linux;Android 14) ExoPlayerLib/2.11.7
Accept: */*
Cookie: __hdnea__=st=1715916110~exp=1716002510~acl=/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/*~hmac=deb7ef23ebf9112155f8055630341f95a7c5f9baa6b27c508d89960d67097d90; __hdnea__=st=1715916110~exp=1716002510~acl=/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/*~hmac=deb7ef23ebf9112155f8055630341f95a7c5f9baa6b27c508d89960d67097d90;
Accept-Charset: UTF-8,*;q=0.8
Accept-Encoding: gzip, deflate
Connection: Keep-Alive
accessToken: (redacted)
appName: RJIL_JioTV
channelid: 291
crmid: (redacted)
deviceId: (redacted)
devicetype: phone
os: android
osversion: 14
srno: 20240516
ssoToken: (redacted)
subscriberid: (redacted)
uniqueid: (redacted)
usergroup: tvYR7NSNn7rymo3F
userid: (redacted)
versionCode: 331
x-platform: android

```

### Behaviour when connecting with local server

Here is the my proxy server response for initial request, which contains cookie (virtually identical to remote server response):

```
HTTP/1.1 200 OK
Connection: Keep-Alive
Content-Encoding: gzip
Content-Length: 3574
x-amz-id-2: XM2fxVrs1S+0H/O2BFvy6MmmSjVOLcSmV0EqB2vKIzxxahqrTcClSYJI3gmaCYyZ752DH/5D2Dc=
x-amz-request-id: CWDHN0R93EF1QB5Y
Last-Modified: Tue, 31 Oct 2023 16:53:03 GMT
ETag: "2ae53a921d2115602f403165c4068626"
x-amz-server-side-encryption: AES256
Content-Type: binary/octet-stream
Server: SimpleHTTP/0.6 Python/3.8.15
Accept-Ranges: bytes
Date: Fri, 17 May 2024 03:53:14 GMT
Via: 1.1 varnish
Age: 35584
X-Served-By: cache-rj-vabb4300026-RJ-VABB
X-Cache: HIT
X-Cache-Hits: 96
X-Timer: S1715917984.770929,VS0,VE0
Vary: Accept-Encoding
Access-Control-Allow-Methods: GET,HEAD,OPTIONS
Access-Control-Allow-Credentials: true
access-control-max-age: 3600
access-control-expose-headers: *
access-control-allow-headers: *
Set-Cookie: __hdnea__=st=1715917983~exp=1716004383~acl=/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/*~hmac=a3ad01e2600415b806a5bd80b6d0c707d3ec354651215a8ded95020080502185; Domain=gec-007-jc.slivcdn.com; path=/; Expires=Sat, 18 May 2024 03:53:03 GMT; SameSite=None; Secure

----------------------  Response unzipped   ----------------------

#EXTM3U
#EXT-X-VERSION:4
#EXT-X-INDEPENDENT-SEGMENTS
#EXT-X-STREAM-INF:BANDWIDTH=389104,AVERAGE-BANDWIDTH=237109,CODECS="avc1.4d400c,mp4a.40.2",RESOLUTION=256x144,FRAME-RATE=25.000,AUDIO="program_audio"
http://127.0.0.1:48996/fixdiscontinuity?manifesturl=https://gec-007-jc.slivcdn.com/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_200k.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=525216,AVERAGE-BANDWIDTH=343045,CODECS="avc1.4d4015,mp4a.40.2",RESOLUTION=426x240,FRAME-RATE=25.000,AUDIO="program_audio"
http://127.0.0.1:48996/fixdiscontinuity?manifesturl=https://gec-007-jc.slivcdn.com/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_300k.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=814736,AVERAGE-BANDWIDTH=526284,CODECS="avc1.77.30,mp4a.40.2",RESOLUTION=640x360,FRAME-RATE=25.000,AUDIO="program_audio"
http://127.0.0.1:48996/fixdiscontinuity?manifesturl=https://gec-007-jc.slivcdn.com/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_500K.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=1192992,AVERAGE-BANDWIDTH=723497,CODECS="avc1.77.30,mp4a.40.2",RESOLUTION=854x480,FRAME-RATE=25.000,AUDIO="program_audio"
http://127.0.0.1:48996/fixdiscontinuity?manifesturl=https://gec-007-jc.slivcdn.com/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_800K.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=1713000,AVERAGE-BANDWIDTH=1133280,CODECS="avc1.4d401f,mp4a.40.2",RESOLUTION=960x540,FRAME-RATE=25.000,AUDIO="program_audio"
http://127.0.0.1:48996/fixdiscontinuity?manifesturl=https://gec-007-jc.slivcdn.com/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_1200K.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=2696240,AVERAGE-BANDWIDTH=1355768,CODECS="avc1.4d401f,mp4a.40.2",RESOLUTION=1280x720,FRAME-RATE=25.000,AUDIO="program_audio"
http://127.0.0.1:48996/fixdiscontinuity?manifesturl=https://gec-007-jc.slivcdn.com/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_2000k.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=3847552,AVERAGE-BANDWIDTH=1783712,CODECS="avc1.4d401f,mp4a.40.2",RESOLUTION=1280x720,FRAME-RATE=25.000,AUDIO="program_audio"
http://127.0.0.1:48996/fixdiscontinuity?manifesturl=https://gec-007-jc.slivcdn.com/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_3000k.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=5452696,AVERAGE-BANDWIDTH=3921845,CODECS="avc1.4d4028,mp4a.40.2",RESOLUTION=1920x1080,FRAME-RATE=25.000,AUDIO="program_audio"
http://127.0.0.1:48996/fixdiscontinuity?manifesturl=https://gec-007-jc.slivcdn.com/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_4200k.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=7051448,AVERAGE-BANDWIDTH=4278185,CODECS="avc1.4d4028,mp4a.40.2",RESOLUTION=1920x1080,FRAME-RATE=25.000,AUDIO="program_audio"
http://127.0.0.1:48996/fixdiscontinuity?manifesturl=https://gec-007-jc.slivcdn.com/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_5500k.m3u8
#EXT-X-MEDIA:TYPE=AUDIO,GROUP-ID="program_audio",LANGUAGE="hin",NAME="Hindi",AUTOSELECT=YES,DEFAULT=YES,URI="1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_Hindi96k.m3u8"
```

ISA dos not make next request correctly as it does not include Cookie from previous response:

```
GET http://127.0.0.1:48996/fixdiscontinuity?manifesturl=https://gec-007-jc.slivcdn.com/HLS/2b54661b-b641-4e40-8bf7-59088ecb31cf/1556784378_show_set_crime_patrol_satark_ep915_episodes_29042018-40.60595104206162_5500k.m3u8 HTTP/1.1
Host: 127.0.0.1:48996
User-Agent: plaYtv/7.1.3 (Linux;Android 14) ExoPlayerLib/2.11.7
Accept: */*
Connection: Keep-Alive
Accept-Charset: UTF-8,*;q=0.8
Accept-Encoding: gzip, deflate
Connection: Keep-Alive
accessToken: (redacted)
appName: RJIL_JioTV
channelid: 291
crmid: (redacted)
deviceId: (redacted)
devicetype: phone
os: android
osversion: 14
srno: 20240516
ssoToken: (redacted)
subscriberid: (redacted)
uniqueid: (redacted)
usergroup: tvYR7NSNn7rymo3F
userid: (redacted)
versionCode: 331
x-platform: android
```

Here is my InputStream config:

```
{
    "label": plugin._title,
    "callback": manifestUrl + "|verifypeer=false",
    "properties": {
        "IsPlayable": True,
        "inputstream": "inputstream.adaptive",
        "inputstream.adaptive.internal_cookies": True,
        "inputstream.adaptive.stream_selection_type": "fixed-res",
        "inputstream.adaptive.chooser_resolution_secure_max": "1080p",
        "inputstream.adaptive.chooser_resolution_max": "1080p",
        "inputstream.adaptive.manifest_headers": urlencode(manifestHeaders),
        "inputstream.adaptive.stream_headers": urlencode(manifestHeaders),
}
```

where `manifestUrl` and `manifestHeaders` are:

```
manifestUrl = "http://127.0.0.1:48996/sonyjiomanifest?channel_id={0}&showtime={1}&srno={2}&programId={3}&begin={4}&end={5}"\
    .format(channel_id, showtime, srno, programId, begin, end)
manifestHeaders = {
   "deviceId":"(redacted)",
   "devicetype":"phone",
   "os":"android",
   "osversion":"14",
   "User-Agent":"plaYtv/7.1.3 (Linux;Android 14) ExoPlayerLib/2.11.7",
   "usergroup":"tvYR7NSNn7rymo3F",
   "versionCode":"331",
   "accessToken":"(redacted)",
   "ssoToken":"(redacted)",
   "crmid":"(redacted)",
   "uniqueid":"(redacted)",
   "userid":"(redacted)",
   "subscriberid":"(redacted)",
   "Accept-Encoding":"gzip, deflate",
   "appName":"RJIL_JioTV",
   "channelid":"291",
   "Connection":"Keep-Alive",
   "srno":"20240516",
   "x-platform":"android"
}
```

## Expected Behavior
ISA should attach cookies always, including for local HTTP server.

## Possible Fix
My theory is that it attaches cookies when making requests to remote HTTPS server, but not when making requests to local HTTP server (`http://127.0.0.1:port/path.....`).

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name: Windows 10
 - Kodi version: Nightly build KodiSetup-20240429-625be876-Omega-x64
 - inputstream.adaptive version:

*note: Once the issue is made we require you to update it with new information should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by kashaCode** [comment]
For example, in CurlUtils.cpp you have code like this that assumes domain.com format where local urls have `http://127.0.0.1:portnumber` format or `http://localhost:portnumber` format:

```
  // Get domain, following format: .example.com
  std::string domain = hostname;
  const size_t dotPos = domain.find('.');
  if (dotPos != std::string::npos)
    domain.erase(0, dotPos);
```

**Comment by CastagnaIT** [comment]
if you replace all manifest urls to use your http server proxy
then you dont neeed to pass cookies to ISA/Kodi its superfluous

your server should manage persistent http sessions
python already do it if you builded a server in right way

anyway about "domain" code mentioned, if you have read the source code note the cookies code on CurlUtils.cpp its an hack,
so you also have to consider the fact that this code in the future should be removed

you should also do a test by disabling internal cookie management on ISA
so that it will be handled by Kodi, done that, the first HTTP requests should result with the cookies, and then later at some point be lost (due to kodi not persistent behaviour)
since curl should work in a similar way of current ISA code, you shoud have same (problem) behaviour

**Comment by kashaCode** [comment]
Sorry for the delay in responding. I did test by setting `"inputstream.adaptive.internal_cookies": False`, and the behaviour remains the same.

Also, thanks for your idea. I had thought handling cookies myself would require a lot of extra work, but requests can handle sessions based on what I looked up. I am new to programming and dealing with HTTP, so I have no idea about best practices. If it works that way I will update this comment.
